src/renderers/WebGLRenderer.js

```diff
@@ -1483,12 +1762,80 @@ function WebGLRenderer( parameters ) {
 
 			} else {
 
-				_this.renderBufferDirect( camera, scene.fog, geometry, material, object, group );
+				var instanceCount = undefined;
+
+				if ( geometry && geometry.isInstancedBufferGeometry && geometry.maxInstancedCount > 0 ) {
+
+					instanceCount = geometry.maxInstancedCount;
+
+				}
+
+				_this.renderBufferDirect( camera, scene.fog, geometry, material, object, group, instanceCount );
+
+			}
+
+			if ( object.onAfterRender ) {
+
+				object.onAfterRender( _this, scene, camera, geometry, material, group );
 
 			}
 
-			object.onAfterRender( _this, scene, camera, geometry, material, group );
+		}
+
+	}
+
+	function renderInstancedGeometryMap ( instancedGeometryMap, scene, camera, overrideMaterial ) {
+
+		for ( var uuid in instancedGeometryMap ) {
+
+			if ( instancedGeometryMap.hasOwnProperty( uuid ) ) {
+
+				var instancedGeometry = instancedGeometryMap[ uuid ];
+				var geometry = instancedGeometry.geometry;
+
+				for (var groupIndex = 0, groupCount = instancedGeometry.groups.length; groupIndex < groupCount; groupIndex++ ) {
+
+					var instancingGroup = instancedGeometry.groups[ groupIndex ];
+
+					var objects = instancingGroup.objects;
+					var material = overrideMaterial ? overrideMaterial : instancingGroup.material;
+
+					if ( objects.length > 0 ) {
+
+						var group = geometry.groups[ groupIndex ] || null;
 
+						for ( var i = 0, l = objects.length; i < l; i++ ) {
+
+							var object = objects[ i ];
+
+							object.modelViewMatrix.multiplyMatrices( camera.matrixWorldInverse, object.matrixWorld );
+							object.normalMatrix.getNormalMatrix( object.modelViewMatrix );
+
+						}
+
+						if ( objects.length < _this.minInstancingBatchSize ) {
+
+							for ( var i = 0, l = objects.length; i < l; i++ ) {
+
+								_this.renderBufferDirect( camera, scene.fog, geometry, material, objects[ i ], group );
```

Comment:
I'm pretty sure that `groupIndex` being null in this context is pre-existing behavior, and intentional. If there are not multiple materials involved, it makes sense to ignore the groups entirely and push a single renderItem for the entire geometry rather than a renderItem per group. My patch above just makes this behave like the old non-instancing code path in that regard.

Your explanation for the keying makes sense, that was my hunch and it's good to have it confirmed. Thanks.